donjuan/randomizer.py

```diff
@@ -41,3 +52,129 @@ def randomize_grid(self, grid: Grid) -> None:
             for j in range(grid.n_cols):
                 self.randomize_cell(grid.cells[i][j])
         return
+
+
+class RoomRandomizer(Randomizer):
+    """
+    Randomize the features of a `Room`.
+    """
+
+    def __init__(
+        self, min_size: int = 2, max_size: int = 4, cell_type: Type[Cell] = SquareCell
+    ):
+        super().__init__()
+        assert issubclass(cell_type, Cell)
+        assert min_size >= 1, f"{min_size}"
+        assert max_size >= min_size, f"{max_size} < {min_size}"
+        assert max_size <= 100, f"{max_size}"  # arbitrary
+        self.min_size = min_size
+        self.max_size = max_size
+        self.cell_type = cell_type
+
+    def randomize_room(self, room: Room) -> None:
+        """
+        Randomly determine the size of the room, and set the cells of
+        the room to a 2D array of unfilled cells of that size.
+        """
+        # Draw the dimensions
+        height = random.randint(self.min_size, self.max_size)
+        width = random.randint(self.min_size, self.max_size)
+
+        # Create empty cells and set them in the room
+        cells = [
+            [self.cell_type(filled=False, coordinates=(i, j)) for j in range(height)]
+            for i in range(width)
+        ]
+        room.set_cells(cells)
+        return
+
+
+class DungeonRoomRandomizer(Randomizer):
+    """
+    Randomize a dungeon by first creating rooms and then applying
+    :meth:`RoomRandomizer.room_randomize` to them.
+
+    Args:
+        room_randomizers (Optional[List[RoomRandomizer]]): randomizers to
+            to apply to each room generated by this randomizer
+        max_num_rooms (Optional[int]): maximum number of rooms to draw,
+            if ``None` then default to the :attr:`max_room_attempts`. See
+            :meth:`DungeonRoomRandomizer.get_number_of_rooms` for details.
+        max_room_attempts (int, optional): default is 100. Maximum number of
+            attempts to generate rooms.
+    """
+
+    def __init__(
+        self,
+        room_randomizers: Optional[List[RoomRandomizer]] = None,
+        max_num_rooms: Optional[int] = None,
+        max_room_attempts: int = 100,
+    ):
+        super().__init__()
+        self.room_randomizers = room_randomizers or [RoomRandomizer()]
+        for room_randomizer in self.room_randomizers:
+            assert isinstance(room_randomizer, RoomRandomizer)
+        self.max_num_rooms = max_num_rooms or max_room_attempts
+        self.max_room_attempts = max_room_attempts
+
+    def get_number_of_rooms(self, dungeon_n_rows: int, dungeon_n_cols: int) -> int:
+        """
+        Randomly determine the number of rooms based on the size
+        of the in coming grid or the :attr:`max_num_rooms` attribute,
+        whichever is less.
+
+        Args:
+            dungeon_n_rows (int): number of rows
+            dungeon_n_cols (int): number of columns
+        """
+        # Loop over room randomizers to find the one with the max `max_size`
```

Comment:
Because we want to have randomizers that do potentially different things. For instance we could have one randomizers that proposes room configurations (the one implemented now just does rectangular rooms), then another that randomly re-shapes rooms, and then another randomizer that randomizes the names of rooms, and then one for features etc.

A practical one that I had in mind here was one randomizer that only makes very large rooms, and one randomizer that only makes very small rooms. This was impossible to do in donjon.

Reply:
> A practical one that I had in mind here was one randomizer that only makes very large rooms, and one randomizer that only makes very small rooms. This was impossible to do in donjon.

In this case, those wouldn't both be passed in at the same time though as they'd be exclusive right?

I get the idea of having different randomizers that do different stuff, but I guess I don't yet see why they would be in a list rather than be distinct attributes with their own functions (i.e., self.room_configuration_randomizer, self.room_name_randomizer, etc)

For example ,`randomize_dungeon` seems pretty specific to the "propose room configuration" randomizer
whereas I'd expect something that is iterating through a list of randomizers to handle them generically. Did you have a refactor for that in mind for down the road?

Reply:
Resolving, since this was changed in the PR to this branch.